PBI-01 Create Interactive Quiz

indiradr opened the pull request:
Create and save interactive quiz feature, which includes creating two types of questions: multiple choice and text question

johaneschristian committed:
[RED] create test for email with same case functionality

In the previous implementation, there is a hidden bug as email with same
characters but different case can be used to register 2 different
accounts. This commit tests the validation and controller method that
serves the functionality for COMPANY.

johaneschristian committed:
[GREEN] implement functionalities to prevent duplicated email

Store email in its lowercase form and perform validation to the
normalized version of email.

johaneschristian committed:
Merge branch 'staging' of https://github.com/one-day-intern/one-day-intern-backend into PBI-01-create_assignment

indiradr committed:
[RED] create stub and model for interactive quiz creation process

create models for interactive quiz as well as the components such as multiple choice question and answer options, and text questions

indiradr committed:
[RED] create tests for interactive quiz creation process

indiradr committed:
[GREEN] implement interactive quiz functionalities

indiradr committed:
[CHORES] resolve merge conflicts from pulling from PBI-01-create_assignment branch

indiradr committed:
[CHORES] delete migrations for users and assessment module

indiradr committed:
[CHORES] remake migrations for users and assessment module

indiradr committed:
[CHORES] resolve merge conflicts from merging with staging

indiradr committed:
[REFACTOR] change method to function as per codacy

refactor to_representation method in models.py to function, which breaks the no-self-use (R0201) constraint

johaneschristian commented on assessment/services/assessment.py:
```diff
@@ -15,7 +17,7 @@ def get_assessor_or_raise_exception(user: User):
 
 def validate_assessment_tool(request_data):
-    if not request_data.get('name'):
+    if not request_data.get('assignment_name'):
```
Hello, Indi. I suppose that your commit was overridden by the merging process. I would resolve this after completing the test flow creation.

indiradr replied:
thanks jo, i'll resolve it right now

indiradr committed:
[REFACTOR] change assignment_name to name in validate_assessment_tool

johaneschristian commented on assessment/services/assessment.py:
```diff
@@ -48,3 +50,130 @@ def create_assignment(request_data, user):
+    content = answer.get('content')
+    correct = answer.get('correct')
+
+    answer_option = MultipleChoiceAnswerOption.objects.create(
```
Hello, Indi. For better encapsulation, I believe it is better to place this inside a method in the MultipleChoiceQuestion.

indiradr replied:
Thanks for the advice, Jo

indiradr committed:
[REFACTOR] change assignment_name to name in save_assignment_to_database and save_interactive_quiz_to_database

johaneschristian commented on one_day_intern/exceptions.py:
```diff
@@ -6,6 +6,10 @@ class InvalidAssignmentRegistration(Exception):
     pass
 
 
+class InvalidInteractiveQuizRegistration(Exception):
```
Hello, Indi. To ensure that the exception will be associated with a 400 response, can you please inherit the InvalidRequestException class instead of Exception? Thank you.

indiradr replied:
Okay I'll refactor that jo

indiradr replied:
<img width="555" alt="Screen Shot 2022-10-26 at 15 12 39" src="https://user-images.githubusercontent.com/70441226/197971476-46704f1d-8c69-4f68-a389-929de5bbd44b.png">
Hi, Jo, is this what you meant?

johaneschristian replied:
Yes, Indi. The reason is the fact that we only put InvalidRequestException in the exception_config.py. In addition, it is still part of an invalid request and hence should yield response 400.

indiradr replied:
Okay, thanks for the clarification, I'll go ahead and push the changes :)

indiradr committed:
[REFACTOR] change InvalidInteractiveQuizRegistration and InvalidAssignmentRegistration

refactor so that both exceptions inherit InvalidRequestException to ensure that the exceptions will be associated with a 400 response

indiradr committed:
[REFACTOR] place save_answer_option_to_database function into MultipleChoiceQuestion model

refactor save_answer_option_to_database function placement for better encapsulation

indiradr committed:
[REFACTOR] remove unused import

remove unused import of MultipleChoiceAnswerOption in assessment.py

indiradr committed:
[CHORES] resolve merge conflicts from merging with staging

indiradr committed:
[REFACTOR] add import and remove duplicate function

import InteractiveQuizSerializer in views.py and remove duplicate function serve_google_register_assessee

johaneschristian committed:
Merge branch 'staging' into PBI-01-create_assignment

johaneschristian committed:
Merge branch 'staging' of https://github.com/one-day-intern/one-day-intern-backend into PBI-01-create_assignment

johaneschristian committed:
Merge branch 'PBI-01-create_assignment' of https://github.com/one-day-intern/one-day-intern-backend into PBI-01-create_assignment

johaneschristian committed:
[CHORES] merge and refactor branch staging and PBI-01-create_assignment
